src/devices/bus/centronics/etibuffer.cpp

```diff
@@ -0,0 +1,369 @@
+	: device_t(mconfig, type, tag, owner, clock)
+	, device_centronics_peripheral_interface(mconfig, *this)
+	, m_maincpu(*this, "maincpu")
+	, m_screen(*this, "screen")
+	, m_ctx(*this, ETI_CENTRONICS_TAG)
+	, m_ctx_data_out(*this, "ctx_data_out")
+	, m_printerready_led(*this, "printerready_led")
+	, m_bufferready_led(*this, "bufferready_led")
+{
+}
+
+//--------------------------------------------------
+//  mem map
+//--------------------------------------------------
+
+void etiprintbuffer_device::etiprintbuffer_device_memmap(address_map &map)
+{
+	map.unmap_value_high();
+	map(0x0000, 0x7ff).rom().region("maincpu", 0);
+	map(0x1000, 0x1fff).r(FUNC(etiprintbuffer_device::eti_read_1000));
+	map(0x2000, 0x2fff).w(FUNC(etiprintbuffer_device::eti_write_2000));
+	map(0x3000, 0x3fff).w(FUNC(etiprintbuffer_device::eti_write_3000));
+}
+
+//--------------------------------------------------
+//  io map
+//--------------------------------------------------
+
+void etiprintbuffer_device::etiprintbuffer_device_iomap(address_map &map)
+{
+	map.unmap_value_high();
+	map(0x0, 0xffff).r(FUNC(etiprintbuffer_device::eti_status_r));
+}
+
+//--------------------------------------------------
+//  memory and io functions
+//--------------------------------------------------
+
+uint8_t etiprintbuffer_device::eti_status_r(offs_t offset)
+{
+	return ((m_strobereceived ? 1 : 0 ) << 6) | // d6 is m_strobe status
+			((m_busy          ? 0 : 1 ) << 7);  // d7 is not busy output
+}
+
+uint8_t etiprintbuffer_device::eti_read_1000(offs_t offset)
+{
+	output_ack(0);
+	m_strobereceived = false;
+	m_bufferready_led = !m_strobereceived;
+
+	output_busy(m_strobereceived);
+	m_ack_timer->adjust(attotime::from_usec(ACK_DELAY), TIMER_ACK);
+	return m_datalatch;
+}
+
+void etiprintbuffer_device::eti_write_2000(offs_t offset, uint8_t data)
+{
+	m_ctx_data_out->write(data);
+}
+
+void etiprintbuffer_device::eti_write_3000(offs_t offset, uint8_t data)
+{
+	m_ctx->write_strobe(CLEAR_LINE);
+	m_strobeout_timer->adjust(attotime::from_usec(STROBE_DELAY), TIMER_STROBEOUT);
+}
+
+//--------------------------------------------------
+//  device init
+//--------------------------------------------------
+
+
+void etiprintbuffer_device::device_start()
+{
+	m_printerready_led.resolve();
+	m_bufferready_led.resolve();
+	m_ack_timer = timer_alloc(TIMER_ACK);
+	m_strobeout_timer = timer_alloc(TIMER_STROBEOUT);
+
+	save_item(NAME(m_ram));
+	save_item(NAME(m_data));
+	save_item(NAME(m_datalatch));
+	save_item(NAME(m_strobereceived));
+	save_item(NAME(m_strobe));
+	save_item(NAME(m_busy));
+}
+
+void etiprintbuffer_device::device_reset()
+{
+	address_space &mem = m_maincpu->space(AS_PROGRAM);
+	u16 ram16kblocks = ioport("CONFIG")->read() >> 2;
+	if (ram16kblocks) mem.install_ram(0x4000,0x4000+ram16kblocks*0x4000-1,m_ram);
+
+	output_busy(0);
+	output_fault(1);
+	output_ack(1);
+	output_select(1);
+	output_perror(0);
+	m_ctx->write_strobe(ASSERT_LINE);
```

Comment:
Does the hardware actually have a way of forcing these signal states independently of the CPU?

Reply:
According to the schematic,
only stb, D0-D7, ack and busy are connected on the centronics input.

On the centronics out, it's stb, d0-d7 and busy that are connected.

I set those copying from the printer.cpp because I think the grappler would complain with "NOT SELECTED" if select wasn't set high, for example.

I don't think that output_fault should be 1 though because that would signal a fault?

void centronics_printer_device::device_reset()
{
	m_busy = false;
	output_busy(m_busy);
	output_fault(1);
	output_ack(1);
	output_select(1);
}